crypto/provider_core.c

```diff
@@ -1195,16 +1302,25 @@ int ossl_provider_doall_activated(OSSL_LIB_CTX *ctx,
         if (!CRYPTO_THREAD_write_lock(prov->flag_lock))
             goto err_unlock;
         if (prov->flag_activated) {
-            if (!ossl_provider_up_ref(prov)){
+            /*
+             * We call CRYPTO_UP_REF directly rather than ossl_provider_up_ref
+             * to avoid upping the ref count on the parent provider, which we
+             * must not do while holding locks.
+             */
+            if (CRYPTO_UP_REF(&prov->refcnt, &ref, prov->refcnt_lock) <= 0) {
```

Comment:
It is worth putting this call into its own function and using that here and in ossl_provider_up_ref?  Likewise for the down ref below?

Reply:
I considered it, but didn't think it was worth it for just one simple call like this. I can do if you insist though.

Reply:
I'm not insisting.
Maybe later once it confuses someone :)